ostd/src/arch/x86/cpu/mod.rs

```diff
@@ -385,96 +388,166 @@ cpu_context_impl_getter_setter!(
     [gsbase, set_gsbase]
 );
 
-/// The floating-point state of CPU.
+pub(super) static CPU_FEATURES: Once<FeatureInfo> = Once::new();
+pub(super) static XSTATE_MAX_FEATURES: Once<u64> = Once::new();
+pub(super) static XSAVE_SIZE: Once<usize> = Once::new();
+
+const MAX_FXSAVE_AREA_SIZE: usize = 512;
+
+#[allow(unused)]
+#[repr(packed)]
 #[derive(Clone, Copy, Debug)]
-#[repr(C)]
+struct FxSaveArea {
+    /// 512-byte-long 16-byte-aligned memory region.
+    buffer: [u8; MAX_FXSAVE_AREA_SIZE],
+}
+
+/// Mask features which are restored when returning to user space.
+///
+/// X87 | SSE | AVX | OPMASK | ZMM_HI256 | HI16_ZMM
+const XFEATURE_MASK_USER_RESTORE: u64 = 0b1110_0111;
+
+const MAX_XSAVE_AREA_SIZE: usize = 4096;
+
+#[allow(unused)]
+#[repr(packed)]
+#[derive(Clone, Copy, Debug)]
+struct XSaveArea {
+    legacy: FxSaveArea,
+    features: u64,
+    compaction: u64,
+    reserved: [u64; 6],
+    buffer: [u8; MAX_XSAVE_AREA_SIZE - MAX_FXSAVE_AREA_SIZE - 64],
+}
+
+/// The FPU state save area of user task.
+///
+/// This could be used either for the legacy SSE/MMX FPU state format (as saved
+/// by `FXSAVE` and restored by the `FXRSTOR` instructions), or the modern FPU state
+/// format (as saved by the `XSAVE`` and restored by the `XRSTOR` instructions).
+#[derive(Debug)]
 pub struct FpRegs {
-    buf: FxsaveArea,
-    is_valid: bool,
+    state_area: Segment,
```

Comment:
I can't see why the state area should be represented using untyped memory objects like `Segment`, which completely erases the structure of this area from the type and forces you to mark `XSaveArea` as `#[allow(unused)]`.